File: algorithms/majorityElement/majorityElement.js
/****
 * Solution of https://leetcode.com/problems/majority-element/
 * @author: Jerry Zou
 * @date: 2015.6.4
 **/

/**
 * @param {number[]} nums
 * @return {number}
 */
var majorityElement = function(nums) {
    var hash = {},
        i;

    for (i = 0; i < nums.length; i++) {
        hash[nums[i]] = hash[nums[i]] ? hash[nums[i]]+1 : 1;
    }

    var keys = Object.keys(hash),
        max = 0,
        result;
    for (i = 0; i < keys.length; i++) {
        if (hash[keys[i]] > max) {
            result = keys[i];
            max = hash[keys[i]];
        }
    }

    return parseFloat(result);
};

File: algorithms/cpp/majorityElement/majorityElement.cpp
// Source : https://oj.leetcode.com/problems/majority-element/
// Author : Hao Chen
// Date   : 2014-12-25

/********************************************************************************** 
 * 
 * Given an array of size n, find the majority element. The majority element is the element that appears more than ⌊ n/2 ⌋ times.
 * 
 * You may assume that the array is non-empty and the majority element always exist in the array.
 * 
 * Credits:Special thanks to @ts for adding this problem and creating all test cases.
 *               
 **********************************************************************************/

#include <stdlib.h>
#include <iostream>
#include <vector>
#include <string>
#include <sstream>
using namespace std;


// Moore Voting Algorithm
// Refer to: 
// http://www.cs.utexas.edu/~moore/best-ideas/mjrty/index.html
int majorityElement(vector<int> &num) {
    int majority;
    int cnt = 0;
    for(int i=0; i<num.size(); i++){
        if ( cnt ==0 ){
            majority = num[i];
            cnt++;
        }else{
            majority == num[i] ? cnt++ : cnt --;
            if (cnt >= num.size()/2) return majority;
        }
    }
    return majority;

}


vector<int> &split(const string &s, char delim, vector<int> &elems) {
    stringstream ss(s);
    string item;
    while (getline(ss, item, delim)) {
        elems.push_back(atoi(item.c_str()));
    }
    return elems;
}


vector<int> split(const string &s, char delim) {
    vector<int> elems;
    split(s, delim, elems);
    return elems;
}


int main(int argc, char** argv)
{
    string array = "1,2,1,2,1,2,1,2,1,2,1";
    if (argc > 1){
        array = argv[1];
    }

    cout << "[" << array << "]" << endl;
    vector<int> num = split(array, ',');
    cout << majorityElement(num) <<endl;

    return 0;
}
